httpclient5/src/main/java/org/apache/hc/client5/http/impl/auth/GGSSchemeBase.java

```diff
@@ -60,69 +60,167 @@
  *
  * @since 4.2
  *
- * @deprecated Do not use. The GGS based experimental authentication schemes are no longer
- * supported. Consider using Basic or Bearer authentication with TLS instead.
  */
-@Deprecated
-public abstract class GGSSchemeBase implements AuthScheme {
+// FIXME The class name looks like a Typo. Rename in 6.0 ?
+public abstract class GGSSchemeBase implements AuthSchemeV2 {
 
     enum State {
         UNINITIATED,
-        CHALLENGE_RECEIVED,
-        TOKEN_GENERATED,
+        TOKEN_READY,
+        TOKEN_SENT,
+        SUCCEEDED,
         FAILED,
     }
 
     private static final Logger LOG = LoggerFactory.getLogger(GGSSchemeBase.class);
     private static final String NO_TOKEN = "";
     private static final String KERBEROS_SCHEME = "HTTP";
-    private final org.apache.hc.client5.http.auth.KerberosConfig config;
+
+    // The GSS spec does not specify how long the conversation can be. This should be plenty.
+    // Realistically, we get one initial token, then one maybe one more for mutual authentication.
+    private static final int MAX_GSS_CHALLENGES = 3;
```

Comment:
Why? I don't see a need and have never seen such a boundary. The context is complete when it is complete. In the past 15 years I haven't seen more than one roundtrip for SPNEGO. This can only happen when IAKerb is in place...

Reply:
There is an explicit test to avoid the infinite conversation in the old code, and logic to only ever allow one challenge.

The one challenge case is obviously not working for mutual auth, which relies on processing a second token.

I could argue that this check also protects from malicious DOS attacks.